index.js

```diff
@@ -0,0 +1,57 @@
+const AWS = require('aws-sdk');
+
+/**
+ * @param {object} arc - the parsed .arc file currently executing
+ * @param {object} cloudformation - the current AWS::Serverless CloudFormation template
+ * @param {object} stage - the application stage (one of `staging` or `production`)
+ */
+module.exports = async function arcBucketMacro(arc, cloudformation, stage) {
+  const config = arc.s3;
+  const buckets = {};
+  const aws = new AWS.S3();
+  // for each bucket:
+  await Promise.all(config.map(b => {
+    return new Promise(async (resolve) => {
+      let bucketName = b[0];
+      if (b.length > 1) {
+        // abort if bucket does not match requested stage:
+        if (b[0] !== stage) {
+          return resolve();
+        }
+        // bucketname will be "{name}-production" or "{name}-staging"
+        // if bucket is marked 'unique' it will have a timestamp appended to the end:
+        bucketName = `${b[1]}-${b[0]}${b[2] && b[2].toLowerCase() === 'unique' ? new Date().getTime() : ''}`;
+      }
+      // see if buckets exist on s3:
+      let exists = false;
+      const params = {
+        Bucket: bucketName,
+      };
+      try {
+        const res = await aws.headBucket({ Bucket: bucketName }).promise();
+        exists = true;
+      } catch (e) {
+        // throws error if bucket does not exist
+      }
+      // create buckets if it doesn't exist:
+      if (!exists) {
+        const res = await aws.createBucket({
```

Comment:
what are your thoughts on having cloudformation handle the bucket creation?  that way when we take down the formation, the bucket goes with it

Reply:
It makes sense for arc.code to overwrite the s3 buckets which are part of its deployment, but don't we want manual control over additional buckets?  Maybe this could be an option in the bucket config?

Reply:
what do you mean overwrite the s3 buckets?

Reply:
I'm just saying it might be cleaner to add the s3 bucket creation as part of the cloudformation template instead of doing it separately in code